### ISSUE TITLE
ActiveRecord finders .last and .last(index) behave unexpectedly with offset

### ISSUE THREAD

**Issue by brchristian** [rails/rails#23829]
Given ActiveRecord objects called `Topic` and numbered 1, 2, 3, 4, and 5 for short, I would expect the following:

```
Topic.offset(1).last = 5
Topic.offset(4).last = 5
Topic.offset(1).last(2) = [4, 5]
Topic.offset(4).last(2) = [5]
```

However, the actual behavior in the current Rails `master` is the following:

```
Topic.offset(1).last = 4
Topic.offset(4).last = 1
Topic.offset(1).last(2) = [3, 4]
Topic.offset(4).last(2) = [1]
```

Here are test cases which demonstrate this behavior.

```
def test_last_with_offset
  topic = Topic.order(:id)
  assert_equal topic.to_a.last, topic.last
  assert_equal topic.offset(1).to_a.last, topic.offset(1).last
  assert_equal topic.last, topic.offset(4).last
end

def test_last_with_integer_and_offset
  topic = Topic.order(:id)
  assert_equal topic.to_a.last(2), topic.last(2)
  assert_equal topic.offset(1).to_a.last(2), topic.offset(1).last(2)
  assert_equal topic.last(1), topic.offset(4).last(2)
end
```

**Comment by mdoza** [comment]
@brchristian using the `OFFSET` clause goes along with using the `LIMIT` clause, when the result of `LIMIT + OFFSET` is less than the amount of rows returned, then the first `OFFSET` rows are skipped. You can see this in action, check it out :computer::

When the rows returned are less than `LIMIT + OFFSET`:

```ruby
Topic.offset(1).last
=> #<Topic id: 4>
```
```SQL
SELECT  "topics".* FROM "topics" ORDER BY "topics"."id" DESC LIMIT 1 OFFSET 1
```

So, interestingly enough, to achieve your desired results, if you just call `Topic.offset(1)` alone, `LIMIT` is set to `-1`; rows are omitted at the end of your result:

```ruby
Topic.offset(1)
=> #<ActiveRecord::Relation [#<Topic id: 2>, #<Topic id: 3>, #<Topic id: 4>, #<Topic id: 5>]>
```
```SQL
SELECT "topics".* FROM "topics" LIMIT -1 OFFSET 1
```

Hope this helps :smile:

**Comment by brchristian** [comment]
@mdoza I hear you about how this works in terms of the underlying SQL being generated, but I must admit I find it very unintuitive (and to my knowledge [undocumented](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-last)).

Discussion at https://github.com/rails/rails/issues/23607#issuecomment-186631678 included the philosophy that in general "`relation.last` tries to act in the same way as `relation.to_a.last`".

This seems to violate that intuition.

For instance,

```ruby
Topic.offset(4).to_a
=> [#<Topic id: 5>]

Topic.offset(4).last
=> #<Topic id: 1>
```

This is _very_ surprising to me.

Things appear to get strange with `limit` as well.

```ruby
Topic.offset(4).limit(1).last
=> #<Topic id: 5>

Topic.offset(4).last
=> #<Topic id: 1>
````

Again, this to me seems _very_ surprising.

Maybe I'm an unrepresentative or naive user (which is totally possible). Is this really the behavior users would expect?

**Comment by chashmeetsingh** [comment]
totally agree with @mdoza

**Comment by brchristian** [comment]
In https://github.com/rails/rails/pull/23598#issuecomment-189675440, DHH says that `Topic.offset(1).second_to_last` should return `<Topic id:  4>`. This would imply that `Topic.offset(1).last` should return `<Topic id:  5>`. This behavior also seems sensible to me.

**Comment by brchristian** [comment]
Just responding to rails-bot here to note that the discussion is ongoing between myself, sgrif, prathamesh-sonpatki, bogdan and others at https://github.com/rails/rails/issues/24109.

**Comment by kamipo** [comment]
Fixed by 68754d37706f11ca363b3108eda10749b9957817.
